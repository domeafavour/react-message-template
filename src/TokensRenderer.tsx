import React from "react";
import {
  getNoVariablesMatchedKey,
  getTextTokenKey,
  getVariableTokenKey,
} from "./keys";
import {
  InputTemplateValues,
  NoMatch,
  StringTemplateKeys,
  TemplateToken,
} from "./typings";

export function defaultNoMatch(key: string) {
  return `{${key}}`;
}

interface Props<S extends string = string> {
  tokens: TemplateToken<StringTemplateKeys<S>>[];
  values?: InputTemplateValues<S>;
  noMatch?: NoMatch<S>;
}

export type { Props as TokensRendererProps };

export function TokensRenderer<S extends string = string>({
  tokens,
  values,
  noMatch = defaultNoMatch as NoMatch<S>,
}: Props<S>) {
  return tokens.map((token) => {
    if (token.type === "text") {
      return (
        <React.Fragment key={getTextTokenKey(token)}>
          {token.text}
        </React.Fragment>
      );
    }

    const value = values?.[token.key as keyof InputTemplateValues<S>];

    if (typeof value !== "undefined") {
      return (
        <React.Fragment key={getVariableTokenKey(token)}>
          {value as React.ReactNode}
        </React.Fragment>
      );
    }

    return (
      <React.Fragment key={getNoVariablesMatchedKey(token)}>
        {noMatch(token.key, token)}
      </React.Fragment>
    );
  });
}
