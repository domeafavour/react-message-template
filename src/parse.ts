import {
  StringTemplateKeys,
  TemplateToken,
  TextToken,
  VariableToken,
} from "./typings";

export function createTextToken(
  text: string,
  start: number,
  end: number
): TextToken {
  return { type: "text", start, end, text };
}

export function createVariableToken<K extends string = string>(
  key: K,
  start: number,
  end: number
): VariableToken<K> {
  return { type: "variable", start, end, key };
}

export function tokenizer<S extends string = string>(input: S) {
  const VARIABLE_PATTERN = /{([^}]+)}/g;
  const tokens: TemplateToken<StringTemplateKeys<S>>[] = [];
  let match: RegExpExecArray | null;
  let lastIndex = 0;

  while ((match = VARIABLE_PATTERN.exec(input))) {
    const [full, key] = match;
    const start = match.index;
    const end = start + full.length;
    if (start > lastIndex) {
      tokens.push(
        createTextToken(input.slice(lastIndex, start), lastIndex, start)
      );
    }
    tokens.push(createVariableToken(key as StringTemplateKeys<S>, start, end));
    lastIndex = end;
  }

  if (lastIndex < input.length) {
    tokens.push(
      createTextToken(input.slice(lastIndex), lastIndex, input.length)
    );
  }

  return tokens;
}
