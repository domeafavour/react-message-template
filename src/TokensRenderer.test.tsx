import { render } from "@testing-library/react";
import { describe, expect, it } from "vitest";
import { TokensRenderer } from "./TokensRenderer";
import { createTextToken, createVariableToken } from "./parse";

describe("TokensRenderer", () => {
  it("should render correctly", () => {
    const result = render(
      <TokensRenderer<"Hello, {name}!">
        tokens={[
          createTextToken("Hello, ", 0, 7),
          createVariableToken("name", 7, 13),
          createTextToken("!", 13, 14),
        ]}
        values={{ name: <strong>John Doe</strong> }}
      />
    );
    // <div>Hello, <strong>John Doe</strong>!</div>
    expect(result.container).toMatchSnapshot();
  });
});
