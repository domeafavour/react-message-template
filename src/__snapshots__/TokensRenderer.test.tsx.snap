// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TokensRenderer > should render correctly 1`] = `
<div>
  Hello, 
  <strong>
    John Doe
  </strong>
  !
</div>
`;
